Size CupertinoTextSelectionToolbar to children

tgucio opened the pull request:
Fixes https://github.com/flutter/flutter/issues/133140, https://github.com/flutter/flutter/issues/132819, https://github.com/flutter/flutter/issues/69847

Before             |  After
:-------------------------:|:-------------------------:
![before](https://github.com/flutter/flutter/assets/72562119/3226290f-d0c6-4bb8-a9c2-dc8c3d88c9d8)  |  ![after](https://github.com/flutter/flutter/assets/72562119/19bcb75b-8aae-4aa0-9d00-4719b88eb7a7)


## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

tgucio committed:
Allow toolbar to grow with children

tgucio committed:
const

tgucio committed:
Redundant param

tgucio committed:
Remove default param

tgucio committed:
Fix test

tgucio committed:
Fix some TextField tests

tgucio committed:
Fix child height calc

tgucio committed:
Merge branch 'master' into cupertino-toolbar

tgucio committed:
Use Canvas.drawShadow to draw shadow

tgucio committed:
Use const

tgucio committed:
Bring shadow test back

tgucio committed:
Adjust shadow

tgucio committed:
Shadow test values

tgucio committed:
Use padding on either side

tgucio committed:
Fix tests

tgucio committed:
Remove padding

justinmc commented on packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:
```diff
@@ -14,19 +14,22 @@ import 'text_selection_toolbar_button.dart';
 
 // Values extracted from https://developer.apple.com/design/resources/.
-// The height of the toolbar, including the arrow.
-const double _kToolbarHeight = 45.0;
```
I've always wanted to get rid of the hard coded height if possible :+1:

justinmc commented on packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:
```diff
@@ -14,19 +14,22 @@ import 'text_selection_toolbar_button.dart';
 // screen. Eyeballed value.
 const double _kArrowScreenPadding = 26.0;
 
-// Values extracted from https://developer.apple.com/design/resources/.
```
Can you keep this comment or is there a reason to remove it?

tgucio replied:
The same comment appears above - I separated it by one line so it's clearer it refers to all the constants.

LongCatIsLooong replied:
Many other constants here are eyeballed so I think it will be good to keep the provenance per constant, or per constant group.

tgucio replied:
I'll keep comments per constant.

justinmc reviewed (commented):
Thanks for working on this, it's a big problem for people!

Some things I see in the screenshot in https://github.com/flutter/flutter/pull/133386#issuecomment-1701292684:

  1. The right arrow seems to not be vertically aligned.
  2. The shadow above the toolbar is missing.  I checked and native (iOS 14) does have a shadow above.
  3. The toolbar background looks darker.

What happens if the text in the toolbar is very big, or you use the Thai (I think) text from https://github.com/flutter/flutter/issues/133140?

tgucio committed:
Merge branch 'master' into cupertino-toolbar

tgucio committed:
Paint shadow using BoxShadow and drawRRect

tgucio committed:
Center chevrons vertically

tgucio committed:
Use Align to layout nav buttons

tgucio committed:
Update comment

tgucio committed:
Fix shadow test

tgucio committed:
Improve shadows

tgucio committed:
Further refine shadow

tgucio commented:
@justinmc 

> 1. The right arrow seems to not be vertically aligned.
> 2. The shadow above the toolbar is missing.  I checked and native (iOS 14) does have a shadow above.
> 3. The toolbar background looks darker.
> 
> What happens if the text in the toolbar is very big, or you use the Thai (I think) text from #133140?

1. This should be corrected now: a bit tricky as it requires a re-layout of the nav buttons when children have larger height.
2. Yes, this did look differently. I reverted from using `canvas.drawShadow()` to mimicking what `DecoratedBox()` would draw for a shadow. Should be close enough now.
3. I think this effect was caused by the very little shadow on top. The button colours are defined in the button widget I guess.

I've added before and after screenshots that show both above and below toolbar alighnments.

tgucio committed:
Merge branch 'master' into cupertino-toolbar

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:
```diff
@@ -526,9 +548,13 @@ class _CupertinoTextSelectionToolbarContentState extends State<_CupertinoTextSel
+      // _CupertinoTextSelectionToolbarItems can get the natural size
+      // of the buttons and then expand vertically as needed i.e. if children
+      // have higher height than that.
+      child: Align(
```
nit: use `Center` instead so it's more obvious the child will be centered?

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:
```diff
@@ -14,19 +14,22 @@ import 'text_selection_toolbar_button.dart';
 // is pointing to. The value used here is eyeballed.
 const double _kToolbarContentDistance = 8.0;
+
+// The size of the arrow button to navigate between pages.
```
I thought this is the arrow that `_clipPath` uses to create the arrow that points to the current selection?

tgucio replied:
Ah, you're right there.

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:
```diff
@@ -296,55 +291,61 @@ class _RenderCupertinoTextSelectionToolbarShape extends RenderShiftedBox {
   }
 
+  // Returns the RRect inside which the child is painted.
+  RRect _shapeRRect() {
```
nit:  make this a getter instead like `isAbove`?

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:
```diff
@@ -354,11 +355,31 @@ class _RenderCupertinoTextSelectionToolbarShape extends RenderShiftedBox {
+    final Path clipPath = _clipPath(rrect);
+
+    // If configured, paint the shadow beneath the shape.
+    if (_shadowColor != null) {
```
What's the difference between this and using BoxDecoration?

tgucio replied:
The problem with using `BoxDecoration` to draw shadows was that you had to know upfront whether the toolbar would come above or below the anchor to apply a vertical offset. This worked with const toolbar height but not now we size it to the children.

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:
```diff
@@ -979,6 +1008,21 @@ class _RenderCupertinoTextSelectionToolbarItems extends RenderBox with Container
 
+    // If needed, re-layout slotted children using max. child height to allow
+    // the back/next buttons to be centered vertically.
+    if (greatestHeight > _backButton!.size.height) {
```
Since the menu items are usually text, I think the children here should be baseline-aligned by default? Then this second layout pass is probably no longer needed.

tgucio replied:
IMHO the toolbar should be child type agnostic (e.g. in iOS we already have the LiveText button which uses CustomPaint).

LongCatIsLooong replied:
In that case center aligning sounds like a better option than top aligning the children? Looks nicer when one child is significantly taller.

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:
```diff
@@ -139,21 +147,8 @@ class CupertinoTextSelectionToolbar extends StatelessWidget {
-          ),
-        ],
+      decoration: const BoxDecoration(
+        borderRadius: BorderRadius.all(_kToolbarBorderRadius),
```
Do we still need the border radius here? It will be clipped after the toolbar is painted right?

tgucio replied:
Yeah, this can go - we're not drawing any borders here.

LongCatIsLooong commented:
IIRC in SkParagraph the line height is computed from font metrics instead of glyph (cluster) metrics so stacking diacritics shouldn't make lines taller.

tgucio committed:
Clean up constant comments

tgucio committed:
Use Center instead of Align

tgucio committed:
Make _shapeRRect a getter

tgucio committed:
Make _shapeRRect a method back

tgucio committed:
Merge branch 'master' into cupertino-toolbar

tgucio committed:
Merge branch 'master' into cupertino-toolbar

tgucio committed:
Check child for null in hitTestChildren

tgucio committed:
Dot for double

tgucio committed:
Use child size in _shapeRRect

tgucio committed:
Avoid extra conditions

tgucio committed:
Use const

tgucio committed:
Restore paint method

tgucio committed:
Merge branch 'master' into cupertino-toolbar

tgucio committed:
Remove !

tgucio committed:
Fix tests

tgucio committed:
Restore test offset

tgucio committed:
Merge branch 'master' into cupertino-toolbar

tgucio committed:
Merge branch 'master' into cupertino-toolbar

tgucio committed:
Remove DecoratedBox

tgucio committed:
Center all children vertically

tgucio committed:
Merge branch 'master' into cupertino-toolbar

tgucio commented:
@LongCatIsLooong @justinmc this should be ready for another review.

tgucio committed:
Remove minWidth

tgucio committed:
Update comment

tgucio committed:
Commenet

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:
```diff
@@ -366,26 +360,30 @@ class _RenderCupertinoTextSelectionToolbarShape extends RenderShiftedBox {
-    if (_isAbove) {
+    if (isAbove) {
+      final double arrowBaseY = child.size.height - _kToolbarArrowSize.height;
+      final double arrowTipY = child.size.height;
```
Looks like the path is now described in `child`'s coordinates? If so could you update the comment before this method?

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:
```diff
@@ -395,12 +393,34 @@ class _RenderCupertinoTextSelectionToolbarShape extends RenderShiftedBox {
+        rrect.bottom + _kToolbarArrowSize.height,
+        _kToolbarBorderRadius,
+      ).shift(offset + childParentData.offset + boxShadow.offset);
+      context.canvas.drawRRect(shadowRRect, boxShadow.toPaint());
```
have you tried `Canvas.drawShadow`? According to the documentation it's more efficient.

tgucio replied:
I did and it looked horrible :). Didn't repaint in some cases after tapping buttons.

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:
```diff
@@ -950,8 +988,10 @@ class _RenderCupertinoTextSelectionToolbarItems extends RenderBox with Container
-    final double subsequentPageButtonsWidth =
-        _backButton!.size.width + _nextButton!.size.width;
+    // Require the children to have at least the height of the nav buttons.
+    final double minHeight = clampDouble(constraints.minHeight, _backButton!.size.height, _nextButton!.size.height);
```
Do we still need `minHeight` now that we're vertically centering them?

tgucio replied:
Good point there. I was able to remove this.

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:
```diff
@@ -1021,15 +1059,24 @@ class _RenderCupertinoTextSelectionToolbarItems extends RenderBox with Container
     });
 
+    // Re-layout the children using max. child height to allow them to be centered vertically.
+    visitChildren((RenderObject renderObjectChild) {
```
We should avoid doing a second layout pass if possible. Do these children all have to have the exact same height? To center them vertically could you just adjust their y offsets instead of doing a second layout pass and enforcing `greatestHeight`?

LongCatIsLooong replied:
Also unless `visitChildren` is overridden it doesn't visit the back button and the next button right? Do they need to readjust their heights too?

tgucio replied:
Yes, we need to adjust the children height so the whole button area is painted using another shade during tap - see the screenshot [here](https://github.com/flutter/flutter/pull/133386#discussion_r1331144856) (the bottom toolbar). `visitChildren` also visits the back/next buttons: they're added as render object children (and called "slotted" children) and their heigh is adjusted.

LongCatIsLooong replied:
Ah ok that's unfortunate. Instead of  `RenderObject.layout`, usually `getMaxIntrinsicHeight` is used to do this kind of speculative layout. 

the back/next buttons seem to be added to the parent via `adoptChild`. This doesn't add them to the linked list of children that `visitChildren` traverses.

tgucio replied:
The `visitChildren` method is overriden though:
```dart
  @override
  void visitChildren(RenderObjectVisitor visitor) {
    // Visit the slotted children.
    if (_backButton != null) {
      visitor(_backButton!);
    }
    if (_nextButton != null) {
      visitor(_nextButton!);
    }
    // Visit the list children.
    super.visitChildren(visitor);
  }
```

tgucio commented on packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:
```diff
@@ -252,68 +237,81 @@ class _CupertinoTextSelectionToolbarShape extends SingleChildRenderObjectWidget
-    // top and bottom. Since _kToolbarHeight includes the height of one arrow, the
-    // total height that the child is given is that plus one more arrow height.
-    // The extra height on the opposite side of the arrow will be clipped out. By
+    final BoxConstraints enforcedConstraint = constraints.loosen();
```
I'd say this is something for another PR.

LongCatIsLooong replied:
It was reverted in this patch. Could you add that back if that doesn't break the new layout of the toolbar?

tgucio replied:
Ah, ok - I removed it because the name `heightConstraint` suggested it was height only but indeed it enforced `minWidth` too. Will restore `minWidth`.

tgucio commented on packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:
```diff
@@ -366,26 +360,30 @@ class _RenderCupertinoTextSelectionToolbarShape extends RenderShiftedBox {
+    final BoxParentData childParentData = child.parentData! as BoxParentData;
+    final Offset localAnchor = globalToLocal(isAbove ? _anchorAbove : _anchorBelow);
+
+    final double centerX = childParentData.offset.dx + child.size.width / 2;
```
The 0.0 dx comes from another method (`performLayout()`) so I think it's better to leave it this way in (unlikely) case someone changes something.

tgucio committed:
Remove minHeight for children

tgucio committed:
Merge branch 'master' into cupertino-toolbar

tgucio committed:
Comment

tgucio committed:
Restore arrowTipX with clamp

tgucio commented on packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:
```diff
@@ -366,26 +360,30 @@ class _RenderCupertinoTextSelectionToolbarShape extends RenderShiftedBox {
     }
 
-    final Offset localAnchor = globalToLocal(_anchor);
-    final double arrowTipX = clampDouble(
```
I definitely changed it back by accident. Restoring (and thanks for catching).

tgucio committed:
Use getMaxIntrinsicHeight to determine max child height

tgucio committed:
Restore enforced minWidth

tgucio committed:
Enforce overflowing child minHeight

tgucio committed:
Always initialize hasNextPage/hasPreviousPage in layout

tgucio committed:
Merge branch 'master' into cupertino-toolbar

tgucio committed:
Simplify code

tgucio committed:
Remove assert

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:
```diff
@@ -946,66 +986,71 @@ class _RenderCupertinoTextSelectionToolbarItems extends RenderBox with Container
+          BoxConstraints(
+            maxWidth: firstPageWidth - paginationButtonsWidth,
+            minHeight: greatestHeight,
+            maxHeight: constraints.maxHeight,
```
Should we use a tight height constraint here?

tgucio committed:
Tight height constraint for children

tgucio committed:
Merge branch 'master' into cupertino-toolbar

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/text_selection_toolbar.dart:
```diff
@@ -946,66 +986,71 @@ class _RenderCupertinoTextSelectionToolbarItems extends RenderBox with Container
+    final BoxConstraints slottedConstraints = BoxConstraints(
+      maxWidth: constraints.maxWidth,
+      minHeight: greatestHeight,
+      maxHeight: constraints.maxHeight,
```
Sorry miss this part in the previous review. Should the height constraint here be updated too?

LongCatIsLooong reviewed (approved):
LGTM modulo the height constraints for back/next button. Thanks for fixing this!

tgucio committed:
Tight height constraint for nav buttons

tgucio committed:
Merge branch 'master' into cupertino-toolbar
